src/charm.py

```diff
@@ -77,46 +88,64 @@ def __init__(self, *args):
+    def _on_config_changed(self, _):
         """Handle changed configuration."""
-        self.publish_config()
+        s3_config_data = self._get_s3_storage_config()
+        self.publish_config(s3_config_data)
 
-    def publish_config(self):
+    def _on_mimir_cluster_changed(self, _):
+        self._process_cluster_and_s3_credentials_changes()
+
+    def _on_s3_requirer_credentials_changed(self, _):
+        self._process_cluster_and_s3_credentials_changes()
+
+    def _process_cluster_and_s3_credentials_changes(self):
+        if not self.coordinator.is_coherent():
+            logger.warning("Incoherent deployment: Some required Mimir roles are missing.")
+            return
+        s3_config_data = self._get_s3_storage_config()
+        if s3_config_data == _S3ConfigData() and self.coordinator.has_multiple_workers():
+            logger.warning("Filesystem storage cannot be used with replicated mimir workers")
+            return
+        self.publish_config(s3_config_data)
+
+    def _on_s3_requirer_credentials_gone(self, _):
+        if not self.coordinator.is_coherent():
+            logger.warning("Incoherent deployment: Some required Mimir roles are missing.")
+            return
+        if self.coordinator.has_multiple_workers():
+            logger.warning("Filesystem storage cannot be used with replicated mimir workers")
+            return
+        self.publish_config(_S3ConfigData())
+
+    def publish_config(self, s3_config_data: _S3ConfigData):
         """Generate config file and publish to all workers."""
-        mimir_config = self.coordinator.build_config(dict(self.config))
+        mimir_config = self.coordinator.build_config(s3_config_data)
         self.cluster_provider.publish_configs(mimir_config)
 
-    def _on_mimir_cluster_changed(self, _):
-        if self.coordinator.is_coherent():
-            logger.info("mimir deployment coherent: publishing configs")
-            self.publish_config()
-        else:
-            logger.warning("this mimir deployment is incoherent")
+    def _get_s3_storage_config(self):
+        """Retrieve S3 storage configuration."""
+        if not self.s3_requirer.relations:
+            return _S3ConfigData()
+        raw = self.s3_requirer.get_s3_connection_info()
+        raw["access_key"], raw["secret_key"] = raw.pop("access-key", ""), raw.pop("secret-key", "")
+        return _S3ConfigData(**raw)
 
     def _on_collect_status(self, event: CollectStatusEvent):
         """Handle start event."""
         if not self.coordinator.is_coherent():
             event.add_status(
-                BlockedStatus(
-                    "Incoherent deployment: you are " "lacking some required Mimir roles"
-                )
+                BlockedStatus("Incoherent deployment: you are lacking some required Mimir roles")
+            )
+        s3_config_data = self._get_s3_storage_config()
+        if s3_config_data == _S3ConfigData() and self.coordinator.has_multiple_workers():
```

Comment:
Rather than doing this, I'd have _get_s3_storage_config raise an exception if it fails to validate the databag. It's pydantic after all. The idea is not to always generate a model no matter what, but to only generate a model if the data is valid!